ros/src/pybullet_ros/plugins/control.py

```diff
@@ -102,26 +79,24 @@ def execute(self):
         position_ctrl_task = False
         velocity_ctrl_task = False
         effort_ctrl_task = False
-        for index, subscriber in enumerate(self.pc_subscribers):
-            if subscriber.get_is_data_available():
-                self.position_joint_commands[index] = subscriber.get_last_cmd()
-                position_ctrl_task = True
-        for index, subscriber in enumerate(self.vc_subscribers):
-            if subscriber.get_is_data_available():
-                self.velocity_joint_commands[index] = subscriber.get_last_cmd()
-                velocity_ctrl_task = True
-        for index, subscriber in enumerate(self.ec_subscribers):
-            if subscriber.get_is_data_available():
-                self.effort_joint_commands[index] = subscriber.get_last_cmd()
-                effort_ctrl_task = True
+        if self.pc_subscriber.get_is_data_available():
```

Comment:
Now that the `for` logic is not needed to serialise the individual joint commands, you can consolidate these `if` statements with the ones below.

This would be the compact form:
```python
control_params = {}
if self.pc_subscriber.get_is_data_available():
  control_params["controlMode"] = self.pb.POSITION_CONTROL
  control_params["targetPositions"] = self.pc_subscriber.get_last_cmd()
  control_params["forces] = forces=self.force_commands
elif self.vc_subscriber.get_is_data_available():
  control_params["controlMode"] = self.pb.POSITION_CONTROL
  control_params["targetPositions"] = self.pc_subscriber.get_last_cmd()
  ...
elif self.ec_subscriber.get_is_data_available():
  ...

self.pb.setJointMotorControlArray(bodyUniqueId=self.robot, jointIndices=self.joint_indices, **control_params)
```

But, if it's important to consume the data from the subscriber on every execution (for example, if your message queue is >1 and you don't want velocity or effort commands to stack up in the background), then you can retain this original reading block.

Reply:
I like your version much better. However, I'm not even sure that we will keep it that way. Do you think it makes sense to check for all the topics simultaneously like it is done now or should we rather have one plugin per controller type with a controller manager that decides which command to apply?
If we keep it that way we might have to introduce another flag that indicates which type of command was executed in the step before because in order to apply joint torques, the joints need to be "unlocked" (meaning, you have to send one velocity command with zero velocity before you can apply torques). So I'm not too sure how that would behave if you can just send commands randomly to any topic.

Reply:
Fair question, but for now I think this pattern is better, keeping the interface between the communication and the simulation in one class. As you say, if you have multiple plugins / interfaces then you need a controller manager to arbitrate between them, and that adds more code.

If transitioning between control modes (for example from velocity control to effort control) requires some extra logic, it should be possible to define that here or as part of the message format / protocol.

Reply:
Okay, we will find out whats necessary when I check if effort control works